### DISCUSSION TITLE
Excluding partition from being flashed

### DISCUSSION THREAD

**Discussion by teebr** [OE4T/meta-tegra#1290] [Q&A]
Hello,

I've been looking into having a way to build a standard Yocto image, and then adding a couple of device-specific files at flash time. To do this, I've modified the default partition layout XML file to add a new data partition. 

However, flashing this new partition will generally be a one-off operation. If I re-flash the device later on with an updated image, really I want to flash everything except the new partition. I can't see a straightforward way to do this with NVIDIA's tools. I can see two options, both of which have some issues:
1. `sed` the `flash.xml.in` file to remove the new partition node, and have `doflash.sh` use this reduced file instead. Would that have any impact on what regions of memory are being flashed?
2. read back the existing new partition data, then reflash everything. This seems a bit less janky, though I haven't been able to come up with a single command: variations on `read partition_name partition.data; reboot recovery; x` always seems to fail on x (I had a similar problem when writing a script to flash the kernel and device tree partitions in one ago, though didn't really investigate it).

Any thoughts/ suggestions on the above would be appreciated; I am struggling to get my head around NVIDIA's flashing tools.

Best,

Tom

**Comment by teebr** [comment]
To elaborate a bit further on the second point, it seems that if a write is preceded by a read, then it fails on `tegrarcm_v2` error code 8. For example, this command will work: `read A_kernel kernel.tmp; read A_kernel-dtb dtb.tmp;reboot`, but `read A_kernel kernel.tmp; write A_kernel kernel.tmp;reboot` fails. I've tried a few permutations (adding a reboot-to-recovery between them, adding a sleep), but nothing seems to help. Even splitting them into two separate script calls (i.e. running `doflash.sh` twice with one command each time) fails with the same error. I will try and dig into the flashing tools a bit more to understand what's happening.

**Comment by dwalkes** [comment]
> If I re-flash the device later on with an updated image, really I want to flash everything except the new partition. I can't see a straightforward way to do this with NVIDIA's tools. I can see two options, both of which have some issues:

Another way to do this is to use a scheme like mender or swupdate to update the root partition only, and use this to handle all updates after the first.  You can find a working example of mender updates which use this scheme on the jetpack 4 branches of https://github.com/OE4T/tegra-demo-distro and tegrademo mender.  This approach has an added benefit that it doesn't require physical USB access to peform updates.

> sed the flash.xml.in file to remove the new partition node, and have doflash.sh use this reduced file instead. Would that have any impact on what regions of memory are being flashed?

I expect this should work, also you'll find some old threads here from people who have selectively flashed kernel or devicetree partitions for kernel development, so following those methods might be useful too.

**Reply by teebr** [reply]
Thanks Dan:

> Another way to do this is to use a scheme like mender or swupdate to update the root partition only, and use this to handle all updates after the first.

Eventually I will get to an image-based OTA solution, but still I will need this extra partition for flash-time customisation. For example, as I understand with Mender if you want to pre-authorize devices, you need to deploy a private key to each device, and then upload the corresponding public key to their server. The private key could live on this partition, and would mean I don't have to build a new OS with the private key baked into it for each device.

For production this partition will be a one-off flash and we'll never (...) have to physically reflash, but for development (and pending an image-based OTA solution) it would be really helpful to be able to just leave that partition alone while iterating the image.

> I expect this should work, also you'll find some old threads here from people who have selectively flashed kernel or devicetree partitions for kernel development

This is where I ran into the problem in my comment above: I was hoping to read back the new partition, then flash the whole thing (i.e. re-write the partition data that was just read), but I haven't got to the bottom of why that won't work yet.

**Reply by dwalkes** [reply]
> as I understand with Mender if you want to pre-authorize devices, you need to deploy a private key to each device, and then upload the corresponding public key to their server.

Mender will generate its own I believe, and you can then just authorize it on the server.  So all it needs is a writeable partition.  However I do understand the motivation.

Another way I've done this in the past is by mounting the filesystem before flashing it and writing the custom bits into it on the host.  This works as well.

**Reply by teebr** [reply]
I forgot to mention here, but [NVIDIA has confirmed](https://forums.developer.nvidia.com/t/multiple-tegraflash-commands/260858) that it's not possible to read and then write without performing a HW reset into recovery mode in between. Pending an OTA solution I'm using extlinux so for now can largely get away with just reflashing the rootfs partition when doing updates. Thanks for your input.
